records/generator.go

```diff
@@ -232,24 +258,41 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
 				continue
 			}
 
-			tname := labels.AsDNS952(task.Name)
-			sid := slaveIdTail(task.SlaveId)
-			tag := hashString(task.Id)
-			tail := fname + "." + domain + "."
+			// context used to resolve DomainPattern variables for the current task
+			context := PatternContext{
+				SlaveId: slaveIdTail(task.SlaveId),
+				TaskId: hashString(task.Id),
+				TaskName: labels.AsDNS952(task.Name),
+			}
 
-			// A records for task and task-sid
-			arec := tname + "." + tail
-			rg.insertRR(arec, host, "A")
-			trec := tname + "-" + tag + "-" + sid + "." + tail
+			// task-id A record
+			trec := context.TaskName + "-" + context.TaskId + "-" + context.SlaveId + "." + tail
 			rg.insertRR(trec, host, "A")
 
+			// A records
+			for _, tmpl := range domainTemplates {
+				// skip entries which had an invalid pattern
+				if tmpl == nil {
+					continue
+				}
+
+				// apply pattern to the current record
+				var buf bytes.Buffer
+				err := tmpl.Execute(&buf, context)
+				if err != nil {
+					logging.VeryVerbose.Println("error applying template")
+					continue
+				}
+				rg.insertRR(buf.String()+"."+tail, host, "A")
+			}
+
 			// SRV records
-			if task.Resources.Ports != "" {
+			if task.Resources.Ports != "" && trec != "" {
 				ports := yankPorts(task.Resources.Ports)
 				for _, port := range ports {
 					var srvhost string = trec + ":" + port
-					tcp := "_" + tname + "._tcp." + tail
-					udp := "_" + tname + "._udp." + tail
+					tcp := "_" + context.TaskName + "._tcp." + tail
```

Comment:
The question is whether we expect the users to do a 2-phase service discovery, i.e. querying the task names first with a template driven domain name schema and then the ports based on the result of the first query.

If not, we really should add the template driven, dynamic approach you sketched above.

I tend to prefer the second.